articles/external_qos_file.md

```diff
@@ -0,0 +1,658 @@
+There's no way to match ROS entities to the DDS QoS profile names.
+
+This problem can be solved in two ways:
+- Having conventions to match DDS QoS profiles names to ROS entities.
+  This has been proposed in https://github.com/ros2/rmw_fastrtps/pull/335, and could work well for most use cases.
+- Adding a rmw implementation specific "payload" that it's then used by the rmw implementation.
+  In DDS based implementations, this payload can be just the profile name.
+
+As an example of the second alternative:
+
+```xml
+<qos_profiles>
+    <node name="my_node" namespace="/my_ns/nested_ns">
+        <publisher topic_name="asd">
+            <qos>
+                <rmw_payload>
+                    <profile_name>my_qos_profile_name</profile_name>
+                </rmw_payload>
+            </qos>
+        </publisher>
+    </node>
+</qos_profiles>
+```
+
+```yaml
+/my_ns/nested_ns/my_node:
+    ros__qos_profiles:
+        publisher:
+            topic_name: asd
+            rmw_payload:
+                profile_name: my_qos_profile_name
+```
+
+For this to work, `rmw_qos_profile_t` will need to be extended with a `rmw_qos_profile_payload_t` member defined like:
+
+```c
+typedef struct rmw_qos_profile_payload_item_t {
+  char * key;
+  char * value;
+} rmw_qos_profile_payload_item_t;
+
+typedef struct rmw_qos_profile_payload_t {
+  rmw_qos_profile_payload_item_t * key_value_pairs;
+} rmw_qos_profile_payload_t;
+```
+
+It would be up to the implementation what to do with the passed key value pairs.
+
+Note:
+  The proposed API is limited, if we would want for example to define vendor specific QoS inline in the ROS QoS profile file.
+  If YAML format is used, we could take advantage of it's natural representation of integers,floats,etc, instead of forcing everything to be an string.
+  The complexity of this approach doesn't seem to be worth it.
+Note 2:
+  The rmw specific payload could simply be an string, instead of key value pairs.
+  In that case, the `profile_name` sub-tag would be deleted and `my_qos_profile_name` would be the content of the string.
+
+## File format
+
+As showed in the above examples, both YAML or XML can correctly represent the desired file format.
+
+XML advantages:
+- Can be easily verified.
+
+YAML advantages:
+- It can be argued that the format is simpler and more readable than XML.
+- The format proposed in the above examples can easily be integrated with the current parameter file format.
+  That integration would lead in a unique "configuration file", that could be further extended in the future.
+  For example, passing remapping rules in it.
+
+## Explicitly allowing external configurability in entities
+
+The node options could have a flag to ignore the new `--qos-file` argument.
+In a similar way, there could be an option for publishers/subscriptions/clients and services to allow loading the QoS from an external source (side load).
```

Comment:
I think these two sentences could be refactored, and we could be more explicit about the rationale for this section. E.g.

> Nodes should have an option to enable having their QoS settings configured externally.
> This gives authors the ability to define the allowable range of behavior for their nodes.
> Furthermore, the option to externally configure QoS settings should be disabled by default (ie. node authors must opt-in).
> The rationale for an opt-in policy is to force node authors to have to think about the implications of allowing various QoS changes to the node entities.

Reply:
See https://github.com/ros2/design/pull/296/commits/77cd373b8dc049b63e54e8e926723aef1774d5e4.